File: Scala-ScalaFX/src/main/scala/sevenguis/timer/Timer.scala
package sevenguis.timer

import scalafx.application.JFXApp
import scalafx.application.JFXApp.PrimaryStage
import scalafx.scene.Scene
import scalafx.scene.control._
import scalafx.scene.layout.{VBox, HBox}
import scalafx.geometry.Insets
import scalafx.event.ActionEvent
import scalafx.Includes._
import scalafx.beans.property.DoubleProperty
import javafx.beans.value.ObservableValue
import scalafx.animation.{KeyFrame, Timeline}
import scalafx.util.Duration
import javafx.beans.binding.Bindings
import java.util.concurrent.Callable

object Timer extends JFXApp {
  val progress = new ProgressBar()
  val numericProgress = new Label()
  val slider = new Slider(1, 400, 200)
  val reset = new Button("Reset")

  val elapsed = DoubleProperty(0)
  progress.progress <== elapsed / slider.value
  // Ideally: numericProgress.text <== formatElapsed(elapsed)
  numericProgress.text <== Bindings.createStringBinding(new Callable[String] { override def call(): String =
    formatElapsed(elapsed())
  }, elapsed)
  reset.onAction = (event: ActionEvent) =>  elapsed() = 0

  val timeline = Timeline(KeyFrame(Duration(100), "", (e: ActionEvent) =>
    if (elapsed() < slider.value()) elapsed() = elapsed() + 1))
  timeline.setCycleCount(Timeline.INDEFINITE)
  timeline.play()

  stage = new PrimaryStage {
    title = "Timer"
    scene = new Scene {
      content = new VBox(10) {
        padding = Insets(10)
        content =
          Seq(new HBox(10) { content = Seq(new Label("Elapsed Time: "), progress) },
              numericProgress,
              new HBox(10) { content = Seq(new Label("Duration: "), slider) },
              reset)
      }
    }
  }

  def formatElapsed(elapsed: Double): String = {
    val seconds = (elapsed / 10.0).floor
    val dezipart = elapsed % 10
    seconds.toInt + "." + dezipart.toInt + "s"
  }
}

File: Java8-JavaFX/src/sevenguis/timer/Timer.java
package sevenguis.timer;

import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.application.Application;
import javafx.beans.binding.Bindings;
import javafx.beans.property.SimpleDoubleProperty;
import javafx.geometry.Insets;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.ProgressBar;
import javafx.scene.control.Slider;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;
import javafx.util.Duration;

public class Timer extends Application {

    public void start(Stage stage) {
        ProgressBar progress = new ProgressBar();
        Label numericProgress = new Label();
        Slider slider = new Slider(1, 400, 200);
        Button reset = new Button("Reset");

        SimpleDoubleProperty elapsed = new SimpleDoubleProperty(0);
        progress.progressProperty().bind(elapsed.divide(slider.valueProperty()));
        numericProgress.textProperty().bind(Bindings.createStringBinding(() ->
            formatElapsed(elapsed.intValue()), elapsed));
        // Callback based approach is shorter, ideally the opposite would be the case.
//        elapsed.addListener((v, o, n) -> numericProgress.setText(formatElapsed(n.intValue())));
        reset.setOnAction(event -> elapsed.set(0));

        Timeline timeline = new Timeline(new KeyFrame(Duration.millis(100), event -> {
            if (elapsed.get() < slider.valueProperty().get()) elapsed.set(elapsed.get() + 1);
        }));
        timeline.setCycleCount(Timeline.INDEFINITE);
        timeline.play();

        VBox root = new VBox(10, new HBox(10, new Label("Elapsed Time: "), progress),
                                 numericProgress,
                                 new HBox(10, new Label("Duration: "), slider),
                                 reset);
        root.setPadding(new Insets(10));

        stage.setScene(new Scene(root));
        stage.setTitle("Timer");
        stage.show();
    }

    private static String formatElapsed(int elapsed) {
        int seconds = (int) Math.floor(elapsed / 10.0);
        int dezipart = elapsed % 10;
        return seconds + "." + dezipart + "s";
    }

    public static void main(String[] args) {
        launch(args);
    }
}
